src/charm.py

```diff
@@ -100,23 +98,47 @@ def _on_config_changed(self, event):
 
     def publish_config(self):
         """Generate config file and publish to all workers."""
-        mimir_config = self.coordinator.build_config(dict(self.config))
+        mimir_config = self.coordinator.build_config(self._s3_storage_data)
```

Comment:
this is a race condition waiting to happen. Instead of 'persisting' the data in an instance attr, why don't we make `_s3_connection_info` **return** `S3StorageData` and pass it directly into build_config here?